### PR TITLE
Emoji reacts

### PR DESCRIPTION
**Pull request by RyanRory**
- Sending & receiving emoji reacts.
- Emoji reacts UI under message bubble.
- Emoji reacts list in group chats.
- Quick react & quick cancel.
- Most recent 6 emojis panel shows after long press a message.
- Emoji picker UI.

### CODE REVIEW

**Review by mpretty-cyro — CHANGES_REQUESTED** [review]

File: SessionMessagingKit/Messages/Message.swift [hunk 1]
```diff
@@ -348,6 +349,57 @@ public extension Message {
...
+                        authorId: reactor,
+                        emoji: emoji,
+                        count: count,
+                        sortId: sortId
+                    )
+                    count = 0 // Only insert the first reaction with the total count of this emoji
+                    results.append(reaction)
+                }
+                if rawReaction.you && !reactors.contains(userPublicKey) {
+                    let reaction = Reaction(
```
[the remark this answers is not in the document]
Comment (RyanRory) [reply SessionMessagingKit/Messages/Message.swift]: FYI `.suffix(from: 1)` actually throws...

File: SessionMessagingKit/Open Groups/OpenGroupAPI.swift [hunk 2]
```diff
@@ -618,7 +620,8 @@ public enum OpenGroupAPI {
                 db,
                 request: Request<NoBody, Endpoint>(
                     server: server,
-                    endpoint: .roomMessagesSince(roomToken, seqNo: seqNo)
+                    endpoint: .roomMessagesSince(roomToken, seqNo: seqNo),
+                    queryParameters: [.reactions : "r"]
```
[the remark this answers is not in the document]
Comment (RyanRory) [reply SessionMessagingKit/Open Groups/OpenGroupAPI.swift]: Yep, added.

File: SessionMessagingKit/Open Groups/Models/SOGSMessage.swift [hunk 3]
```diff
@@ -79,15 +98,29 @@ extension OpenGroupAPI.Message {
         self = OpenGroupAPI.Message(
             id: try container.decode(Int64.self, forKey: .id),
             sender: try? container.decode(String.self, forKey: .sender),
-            posted: try container.decode(TimeInterval.self, forKey: .posted),
+            posted: ((try? container.decode(TimeInterval.self, forKey: .posted)) ?? Date().timeIntervalSince1970), // Reaction updates don't include posted
```
[the remark this answers is not in the document]
Comment (RyanRory) [reply SessionMessagingKit/Open Groups/Models/SOGSMessage.swift]: Yep, changed the `posted` to be optional.

File: SessionMessagingKit/Common Networking/QueryParam.swift [hunk 4]
```diff
@@ -9,4 +9,7 @@ enum QueryParam: String {
     case required = "required"
     case limit                      // For messages - number between 1 and 256 (default is 100)
     case platform                   // For file server session version check
+    
+    case reactions = "t"
```
[the remark this answers is not in the document]
Comment (RyanRory) [reply SessionMessagingKit/Common Networking/QueryParam.swift]: Yep, removed.

File: SessionMessagingKit/Open Groups/OpenGroupManager.swift [hunk 5]
```diff
@@ -515,48 +515,65 @@ public final class OpenGroupManager: NSObject {
         
         // Process the messages
         sortedMessages.forEach { message in
-            guard
-                let base64EncodedString: String = message.base64EncodedData,
-                let data = Data(base64Encoded: base64EncodedString)
-            else {
-                // FIXME: Once the SOGS Emoji Reacts update is live we should remove this line (deprecated by the `deleted` flag)
-                messageServerIdsToRemove.append(Int64(message.id))
-                return
```
[the remark this answers is not in the document]
Comment (RyanRory) [reply SessionMessagingKit/Open Groups/OpenGroupManager.swift:524]: This is a bit tricky. I've updated the code to remain this logic only when both `message.base64EncodedData` and `message.reactions` are nil, we remove the message.

[the remark this answers is not in the document]
Comment (mpretty-cyro) [reply SessionMessagingKit/Open Groups/OpenGroupManager.swift]: No worries, I think that should be fine - I expect we'll be doing another release around the start of September where we can properly remove it again (I just wanted a little bit more time than the 1-2 weeks SOGS will have to update before Emoji Reacts gets released)

File: SessionMessagingKit/Sending & Receiving/MessageSender.swift [hunk 6]
```diff
@@ -433,15 +433,16 @@ public final class MessageSender {
             )
             .done(on: DispatchQueue.global(qos: .default)) { responseInfo, data in
                 message.openGroupServerMessageId = UInt64(data.id)
-
+                let serverTimestampMs: UInt64? = (data.posted == nil) ? nil : UInt64(floor(data.posted! * 1000))
```
Comment (mpretty-cyro) [inline SessionMessagingKit/Sending & Receiving/MessageSender.swift:436]: So you should be able to clean this up (and avoid the force-unwrap) by mapping from the nullable type:
```
let serverTimestampMs: UInt64? = data.posted.map { UInt64(floor($0 * 1000)) }
```

**Review by mpretty-cyro — CHANGES_REQUESTED** [review]

**Review by mpretty-cyro — APPROVED** [review]

**Review by mpretty-cyro — APPROVED** [review]

**Pull request merged**